src/server/gc_worker/gc_worker.rs

```diff
@@ -422,6 +454,139 @@ where
         Ok((handled_keys, wasted_keys))
     }
 
+    fn raw_gc_keys(
+        &mut self,
+        keys: Vec<Key>,
+        safe_point: TimeStamp,
+        regions_provider: Option<(u64, Arc<dyn RegionInfoProvider>)>,
+    ) -> Result<(usize, usize)> {
+        let first_key = keys.first().unwrap();
+        let first_raw_key = first_key.clone().append_ts(TimeStamp::max());
+        let range_start_key = first_raw_key.as_encoded();
+
+        let last_key = keys.last().unwrap();
+        let last_raw_key = last_key.clone().append_ts(TimeStamp::zero());
+        let range_end_key = last_raw_key.as_encoded();
+
+        //--------------------------------KeysInRegions-----------------------------
+        struct KeysInRegions<R: Iterator<Item = Region>> {
```

Comment:
It seems the code related to the struct `KeysInRegions` in `raw_gc_keys` and `gc_keys` is totally the same. Instead of copying it here, I suggest you to move it out from the function `gc_keys` so that the two functions can reuse it.